File: server/src/models/Message.ts
import mongoose, { Document, Schema, Types } from 'mongoose';

export type MessageType =
  | 'text'
  | 'image'
  | 'video'
  | 'audio'
  | 'voice'
  | 'document'
  | 'location'
  | 'contact'
  | 'system';

export interface IReaction {
  user: Types.ObjectId;
  emoji: string;
  createdAt: Date;
}

export interface IReadReceipt {
  user: Types.ObjectId;
  readAt: Date;
}

export interface ILocationData {
  latitude: number;
  longitude: number;
  name?: string;
  address?: string;
}

export interface IContactData {
  name: string;
  email?: string;
  username?: string;
  userId?: Types.ObjectId;
}

export interface IMessage extends Document {
  _id: Types.ObjectId;
  clientMessageId: string;
  conversationId: Types.ObjectId;
  senderId?: Types.ObjectId;
  receiverId?: Types.ObjectId;
  source: 'internal' | 'whatsapp_business';
  messageType: MessageType;
  content: string;
  mediaUrl?: string;
  mediaType?: string;
  fileName?: string;
  fileSize?: number;
  location?: ILocationData;
  contact?: IContactData;
  replyTo?: Types.ObjectId;
  forwardedFrom?: Types.ObjectId;
  reactions: IReaction[];
  isEdited: boolean;
  isDeleted: boolean;
  deletedFor: Types.ObjectId[];
  pinned: boolean;
  starredBy: Types.ObjectId[];
  sentAt: Date;
  deliveredAt?: Date;
  readAt?: Date;
  readBy: IReadReceipt[];
  expiresAt?: Date;
  whatsappMessageId?: string;
  whatsappStatus?: 'sent' | 'delivered' | 'read' | 'failed';
  createdAt: Date;
  updatedAt: Date;
}

const ReactionSchema = new Schema<IReaction>({
  user: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  emoji: { type: String, required: true },
  createdAt: { type: Date, default: Date.now },
}, { _id: false });

const ReadReceiptSchema = new Schema<IReadReceipt>({
  user: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  readAt: { type: Date, default: Date.now },
}, { _id: false });

const LocationSchema = new Schema<ILocationData>({
  latitude: { type: Number, required: true },
  longitude: { type: Number, required: true },
  name: { type: String },
  address: { type: String },
}, { _id: false });

const ContactSchema = new Schema<IContactData>({
  name: { type: String, required: true },
  email: { type: String },
  username: { type: String },
  userId: { type: Schema.Types.ObjectId, ref: 'User' },
}, { _id: false });

const MessageSchema = new Schema<IMessage>({
  clientMessageId: { type: String, required: true, index: true },
  conversationId: { type: Schema.Types.ObjectId, ref: 'Conversation', required: true, index: true },
  senderId: { type: Schema.Types.ObjectId, ref: 'User' },
  receiverId: { type: Schema.Types.ObjectId, ref: 'User' },
  source: { type: String, enum: ['internal', 'whatsapp_business'], default: 'internal' },
  messageType: {
    type: String,
    enum: ['text', 'image', 'video', 'audio', 'voice', 'document', 'location', 'contact', 'system'],
    default: 'text',
  },
  content: { type: String, default: '' },
  mediaUrl: { type: String },
  mediaType: { type: String },
  fileName: { type: String },
  fileSize: { type: Number },
  location: { type: LocationSchema },
  contact: { type: ContactSchema },
  replyTo: { type: Schema.Types.ObjectId, ref: 'Message' },
  forwardedFrom: { type: Schema.Types.ObjectId, ref: 'User' },
  reactions: [ReactionSchema],
  isEdited: { type: Boolean, default: false },
  isDeleted: { type: Boolean, default: false },
  deletedFor: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  pinned: { type: Boolean, default: false },
  starredBy: [{ type: Schema.Types.ObjectId, ref: 'User' }],
  sentAt: { type: Date, default: Date.now },
  deliveredAt: { type: Date },
  readAt: { type: Date },
  readBy: [ReadReceiptSchema],
  expiresAt: { type: Date },
  whatsappMessageId: { type: String, index: true },
  whatsappStatus: { type: String, enum: ['sent', 'delivered', 'read', 'failed'] },
}, {
  timestamps: true,
  toJSON: {
    transform(_doc, ret) {
      delete (ret as any).__v;
      return ret;
    }
  }
});

// Index for conversation pagination by createdAt descending
MessageSchema.index({ conversationId: 1, createdAt: -1 });
// TTL index for automatic deletion of disappearing messages
MessageSchema.index({ expiresAt: 1 }, { expireAfterSeconds: 0 });
// Compound index for clientMessageId deduplication per conversation
MessageSchema.index({ conversationId: 1, clientMessageId: 1 }, { unique: true, sparse: true });

export const Message = mongoose.model<IMessage>('Message', MessageSchema);

File: server/src/main/java/com/mchat/model/Message.java
package com.mchat.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.Instant;
import java.util.*;

@Entity
@Table(name = "messages", indexes = {
        @Index(name = "idx_conv_created", columnList = "conversationId, createdAt")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Message {

    @Id
    @Column(name = "id", length = 36, nullable = false, updatable = false)
    @JsonProperty("_id")
    private String id;

    @Column(nullable = false)
    private String clientMessageId;

    @Column(nullable = false, length = 36)
    private String conversationId;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "sender_id")
    @JsonProperty("senderId")
    private User sender;

    private String receiverId;

    @Builder.Default
    private String source = "internal"; // internal, whatsapp_business

    @Column(nullable = false, length = 20)
    @Builder.Default
    private String messageType = "text"; // text, image, video, audio, voice, document, location, contact, system

    @Column(length = 5000, columnDefinition = "TEXT")
    private String content;

    @Column(length = 1000)
    private String mediaUrl;

    private String mediaType;
    private String fileName;
    private Long fileSize;

    // Location details
    private Double locationLat;
    private Double locationLng;
    private String locationName;
    private String locationAddress;

    // Contact details
    private String contactName;
    private String contactEmail;
    private String contactUsername;
    private String contactUserId;

    // Reply and forwarded
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "reply_to_id")
    private Message replyTo;

    private String forwardedFrom;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "message_reactions", joinColumns = @JoinColumn(name = "message_id"))
    @Builder.Default
    private List<ReactionItem> reactions = new ArrayList<>();

    @Builder.Default
    private Boolean isEdited = false;

    @Builder.Default
    private Boolean isDeleted = false;

    @Builder.Default
    private Boolean pinned = false;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "message_starred", joinColumns = @JoinColumn(name = "message_id"))
    @Column(name = "user_id")
    @Builder.Default
    private Set<String> starredBy = new HashSet<>();

    @Builder.Default
    private String status = "sent"; // sending, sent, delivered, read, failed

    private Instant sentAt;
    private Instant deliveredAt;
    private Instant readAt;
    private Instant expiresAt;

    private String whatsappMessageId;
    private String whatsappStatus;

    @CreationTimestamp
    private Instant createdAt;

    @UpdateTimestamp
    private Instant updatedAt;

    @PrePersist
    public void prePersist() {
        if (this.id == null || this.id.isBlank()) {
            this.id = UUID.randomUUID().toString();
        }
        if (this.sentAt == null) {
            this.sentAt = Instant.now();
        }
    }

    @JsonProperty("location")
    public Map<String, Object> getLocation() {
        if (locationLat == null || locationLng == null) return null;
        Map<String, Object> loc = new HashMap<>();
        loc.put("latitude", locationLat);
        loc.put("longitude", locationLng);
        loc.put("name", locationName);
        loc.put("address", locationAddress);
        return loc;
    }

    @JsonProperty("contact")
    public Map<String, Object> getContact() {
        if (contactName == null) return null;
        Map<String, Object> c = new HashMap<>();
        c.put("name", contactName);
        c.put("email", contactEmail);
        c.put("username", contactUsername);
        c.put("userId", contactUserId);
        return c;
    }

    @Embeddable
    @Getter
    @Setter
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class ReactionItem {
        @Column(name = "user_id")
        @JsonProperty("user")
        private String user;
        private String emoji;
        @Builder.Default
        private Instant createdAt = Instant.now();
    }
}
